# Pandas

## What is panadas

- Pandas is a Python library for data analysis and manipulation.
- It is built on top of the NumPy library.
- It offers data structures and operations for manipulating numerical tables and time series.
- It provides data analysis tools for exploratory data analysis, data cleaning, and data transformation.
- It also provides data visualization features.
- It's free to use and open source.

## Installation

To install pandas, you can use the following command:

```bash
conda install -c conda-forge pandas
```

## Import pandas

To import pandas, you can use the following command:

In [1]:
import pandas as pd

**Note :** When importing pandas, it's by convention to use the alias `pd`.

## Pandas - Data Structures

Pandas has two main data structures: `Series` and `DataFrames`.

- A DataFrame is a two-dimensional labeled data structure with columns of potentially different types.
- A Series is a one-dimensional labeled array.

## Pandas - Data Structures
### DataFrames

A DataFrame is a table. It contains an array of individual entries, each of which has a certain value. Each entry corresponds to a row (or record) and a column

In [7]:
a =pd.DataFrame({'Yes': [50, 21], 'No': [131, 2]})
print(a)
a

   Yes   No
0   50  131
1   21    2


Yes   No
0   50  131
1   21    2

In the previous example:
  - `Yes` and `No` are the column names
  - `50` and `21` are the values in the `Yes` column
  - `131` and `2` are the values in the `No` column
  - yes/0 is 50; yes/1 is 21; no/0 is 131; no/1 is 2

### DataFrames

The `Dataframes` object is not limited to numbers values. It can contain any data type.

In [12]:
df = pd.DataFrame({'Bob': ['I liked it.', 'It was awful.'], 'Sue': ['Pretty good.', 'Bland.']})
print(df)
df["Bob"][0]

             Bob           Sue
0    I liked it.  Pretty good.
1  It was awful.        Bland.


'I liked it.'

We are using the `pd.DataFrame()` constructor to instaciate these DataFrame objects. 

The syntax for declaring a new one is a dictionary whose keys are the column names (`Bob` and `Sue` in this example), and whose values are a list of entries. This is the standard way of constructing a new DataFrame, and the one you are most likely to encounter.

The dictionary-list constructor assigns values to the column labels, but just uses an *ascending count* from 0 (0, 1, 2, 3, ...) for the row labels. Sometimes this is OK, but oftentimes we will want to assign these labels ourselves.

The list of row labels used in a DataFrame is known as an `Index`. We can assign values to it by using an index parameter in our constructor:

In [20]:
df = pd.DataFrame({
    'Bob': ['I liked it.', 'It was awful.'], 
    'Sue': ['Pretty good.', 'Bland.']
}, index=['Product A', 'Product B'])
df["Bob"]["Product A"]

'I liked it.'

In [19]:
pd.DataFrame({'Bob': ['I liked it.', 'It was awful.', 2], 
              'Sue': ['', 'Bland.', 1]},
             index=['Product A', 'Product B', "Product_count"])

Bob     Sue
Product A        I liked it.        
Product B      It was awful.  Bland.
Product_count              2       1

## Pandas - Data Structures
### Series

A Series, by contrast, is a sequence of data values. If a DataFrame is a table, a Series is a list :

In [21]:
s = pd.Series([1, 2, 3, 4, 5])
s

0    1
1    2
2    3
3    4
4    5
dtype: int64

**Note :** In fact you can create one with nothing more than a python list

### Series

You can visualize a Series as a single column of a DataFrame. So you can assign row labels to the Series the same way as before, using an `index` parameter. However, a Series does not have a column name, it only has one overall name:

In [27]:
serie_label = pd.Series(["30", 35, 40], index=['2015 Sales', '2016 Sales', '2017 Sales'], name='Product A')
print("serie_label", serie_label, sep="\n")
serie_label["2016 Sales"]

serie_label
2015 Sales    30
2016 Sales    35
2017 Sales    40
Name: Product A, dtype: object


35

## Pandas - Data Structures
### Conclusion

The Series and the DataFrame are intimately related. It's helpful to think of a `DataFrame` as actually being just a bunch of Series "glued together".

## Pandas - Reading Data Files

Being able to create a DataFrame or Series by hand is handy. But, most of the time, we won't actually be creating our own data by hand. Instead, we'll be working with data that already exists.

Pandas has built-in methods for reading data from various file formats including CSV, Excel, JSON, and SQL.

## Pandas - Reading Data Files

Here is a table of the different types of files that pandas can read:

| Data Format | Read | Write |
|------------|------|-------|
| CSV | pd.read_csv | pd.to_csv |
| Excel | pd.read_excel | pd.to_excel |
| JSON | pd.read_json | pd.to_json |
| SQL | pd.read_sql | pd.to_sql |
| HTML | pd.read_html | pd.to_html |


**Note :** in this notebook we'll use the CSV file format

In [33]:
wine_reviews = pd.read_csv("./data/numpy_pandas/dataset/wine-reviews/winemag-data-130k-v2.csv")
wine_reviews.shape

(129971, 14)

**Note :** We can see that this dataset has 129971 rows and 14 columns. This is like 1.8 Millions of entries.

## Pandas - Reading Data Files
### pd.read_csv()

The `pd.read_csv()` method contains many parameters :

| Parameter | Description |
|-----------|-------------|
| filepath_or_buffer | The path to the file to read. Can be a URL. |
| sep | The separator for the columns. By default, it is a comma. |
| delimiter | The separator for the columns. By default, it is a comma. |
| header | The row to use as the column names. By default, it is the first row. |
| names | The column names. By default, it is None. |
| index_col | The column to use as the index. By default, it is None. |
| usecols | The columns to use. By default, it is None. |
| encoding | The encoding to use. By default, it is None. |
| engine | The engine to use. By default, it is None. |
| squeeze | If the data is only one column, return a Series. By default, it is False. |
| skiprows | The rows to skip. By default, it is None. |
| nrows | The number of rows to read. By default, it is None. |
| na_values | The values to use for missing values. By default, it is None. |
| parse_dates | If True, parse the dates. By default, it is False. |
| infer_datetime_format | If True, infer the datetime format. By default, it is False. |

## Pandas - Reading Data Files
### Head and Tail

We can visulize the first and last rows of a DataFrame:
  - If I call directly the `DataFrame` object, I'll see the first 5 rows and the last 5 rows
  - If I call the `head()` method, I'll see the first 5 rows
  - If I call the `tail()` method, I'll see the last 5 rows

In [34]:
(wine_reviews)

Unnamed: 0   country   
0                0     Italy  \
1                1  Portugal   
2                2        US   
3                3        US   
4                4        US   
...            ...       ...   
129966      129966   Germany   
129967      129967        US   
129968      129968    France   
129969      129969    France   
129970      129970    France   

                                              description   
0       Aromas include tropical fruit, broom, brimston...  \
1       This is ripe and fruity, a wine that is smooth...   
2       Tart and snappy, the flavors of lime flesh and...   
3       Pineapple rind, lemon pith and orange blossom ...   
4       Much like the regular bottling from 2012, this...   
...                                                   ...   
129966  Notes of honeysuckle and cantaloupe sweeten th...   
129967  Citation is given as much as a decade of bottl...   
129968  Well-drained gravel soil gives this wine its c...   
129969  A dry style of Pinot Gris, this is crisp with ...   
129970  Big, rich and off-dry, this is powered by inte...   

                                   designation  points  price   
0                                 Vulkà Bianco      87    NaN  \
1                                     Avidagos      60   15.0   
2                                          NaN      87   14.0   
3                         Reserve Late Harvest      87   13.0   
4           Vintner's Reserve Wild Child Block      87   65.0   
...                                        ...     ...    ...   
129966  Brauneberger Juffer-Sonnenuhr Spätlese      90   28.0   
129967                                     NaN      90   75.0   
129968                                   Kritt      90   30.0   
129969                                     NaN      90   32.0   
129970           Lieu-dit Harth Cuvée Caroline      90   21.0   

                 province             region_1           region_2   
0       Sicily & Sardinia                 Etna                NaN  \
1                   Douro                  NaN                NaN   
2                  Oregon    Willamette Valley  Willamette Valley   
3                Michigan  Lake Michigan Shore                NaN   
4                  Oregon    Willamette Valley  Willamette Valley   
...                   ...                  ...                ...   
129966              Mosel                  NaN                NaN   
129967             Oregon               Oregon       Oregon Other   
129968             Alsace               Alsace                NaN   
129969             Alsace               Alsace                NaN   
129970             Alsace               Alsace                NaN   

               taster_name taster_twitter_handle   
0            Kerin O’Keefe          @kerinokeefe  \
1               Roger Voss            @vossroger   
2             Paul Gregutt           @paulgwine    
3       Alexander Peartree                   NaN   
4             Paul Gregutt           @paulgwine    
...                    ...                   ...   
129966  Anna Lee C. Iijima                   NaN   
129967        Paul Gregutt           @paulgwine    
129968          Roger Voss            @vossroger   
129969          Roger Voss            @vossroger   
129970          Roger Voss            @vossroger   

                                                    title         variety   
0                       Nicosia 2013 Vulkà Bianco  (Etna)     White Blend  \
1           Quinta dos Avidagos 2011 Avidagos Red (Douro)  Portuguese Red   
2           Rainstorm 2013 Pinot Gris (Willamette Valley)      Pinot Gris   
3       St. Julian 2013 Reserve Late Harvest Riesling ...        Riesling   
4       Sweet Cheeks 2012 Vintner's Reserve Wild Child...      Pinot Noir   
...                                                   ...             ...   
129966  Dr. H. Thanisch (Erben Müller-Burggraef) 2013 ...        Riesling   
129967                  Citation 2004 P

In [37]:
(wine_reviews.head(20))

Unnamed: 0    country                                        description   
0            0      Italy  Aromas include tropical fruit, broom, brimston...  \
1            1   Portugal  This is ripe and fruity, a wine that is smooth...   
2            2         US  Tart and snappy, the flavors of lime flesh and...   
3            3         US  Pineapple rind, lemon pith and orange blossom ...   
4            4         US  Much like the regular bottling from 2012, this...   
5            5      Spain  Blackberry and raspberry aromas show a typical...   
6            6      Italy  Here's a bright, informal red that opens with ...   
7            7     France  This dry and restrained wine offers spice in p...   
8            8    Germany  Savory dried thyme notes accent sunnier flavor...   
9            9     France  This has great depth of flavor with its fresh ...   
10          10         US  Soft, supple plum envelopes an oaky structure ...   
11          11     France  This is a dry wine, very spicy, with a tight, ...   
12          12         US  Slightly reduced, this wine offers a chalky, t...   
13          13      Italy  This is dominated by oak and oak-driven aromas...   
14          14         US  Building on 150 years and six generations of w...   
15          15    Germany  Zesty orange peels and apple notes abound in t...   
16          16  Argentina  Baked plum, molasses, balsamic vinegar and che...   
17          17  Argentina  Raw black-cherry aromas are direct and simple ...   
18          18      Spain  Desiccated blackberry, leather, charred wood a...   
19          19         US  Red fruit aromas pervade on the nose, with cig...   

                                          designation  points  price   
0                                        Vulkà Bianco      87    NaN  \
1                                            Avidagos      60   15.0   
2                                                 NaN      87   14.0   
3                                Reserve Late Harvest      87   13.0   
4                  Vintner's Reserve Wild Child Block      87   65.0   
5                                        Ars In Vitro      87   15.0   
6                                             Belsito      87   16.0   
7                                                 NaN      87   24.0   
8                                               Shine      87   12.0   
9                                         Les Natures      87   27.0   
10                                     Mountain Cuvée      87   19.0   
11                                                NaN      87   30.0   
12                                                NaN      87   34.0   
13                                              Rosso      87    NaN   
14                                                NaN      87   12.0   
15                                              Devon      87   24.0   
16                                              Felix      87   30.0   
17                                Winemaker Selection      87   13.0   
18  Vendimia Seleccionada Finca Valdelayegua Singl...      87   28.0   
19                                                NaN      87   32.0   

             province             region_1           region_2   
0   Sicily & Sardinia                 Etna                NaN  \
1               Douro                  NaN                NaN   
2              Oregon    Willamette Valley  Willamette Valley   
3            Michigan  Lake Michigan Shore                NaN   
4              Oregon    Willamette Valley  Willamette Valley   
5      Northern Spain              Navarra                NaN   
6   Sicily & Sardinia             Vittoria                NaN   
7              Alsace               Alsace                NaN   
8         Rheinhessen                  NaN                NaN   
9              Alsace               Alsace                NaN   
10         California          Napa Valley               Napa   
11             Alsace         

In [41]:
(wine_reviews.tail(1))

Unnamed: 0 country                                        description   
129970      129970  France  Big, rich and off-dry, this is powered by inte...  \

                          designation  points  price province region_1   
129970  Lieu-dit Harth Cuvée Caroline      90   21.0   Alsace   Alsace  \

       region_2 taster_name taster_twitter_handle   
129970      NaN  Roger Voss            @vossroger  \

                                                    title         variety   
129970  Domaine Schoffit 2012 Lieu-dit Harth Cuvée Car...  Gewürztraminer  \

                  winery  
129970  Domaine Schoffit

If we look at the data more precisely, we can see that the dataset has a built-in index. If we want to use this column as index for the data we need to specify the `index_col` parameter.


In [43]:
wine_reviews = pd.read_csv("./data/numpy_pandas/dataset/wine-reviews/winemag-data-130k-v2.csv", index_col=0)
wine_reviews.shape

(129971, 13)

## Pandas - Indexing, Selecting, Assigning
### Introduction

Selecting specific values of a pandas DataFrame or Series to work on is an implicit step in almost any data operation you'll run, so one of the first things you need to learn in working with data in Python is how to go about selecting the data points relevant to you quickly and effectively.

In [2]:
import pandas as pd
pd.set_option('display.max_rows', 10)
reviews = pd.read_csv("./data/numpy_pandas/dataset/wine-reviews/winemag-data-130k-v2.csv", index_col=0)

## Pandas - Indexing, Selecting, Assigning
### Native accessors

Native Python objects provide good ways of indexing data. Pandas carries all of these over, which helps make it easy to start with.

Consider this DataFrame:

In [3]:
reviews

country                                        description   
0          Italy  Aromas include tropical fruit, broom, brimston...  \
1       Portugal  This is ripe and fruity, a wine that is smooth...   
2             US  Tart and snappy, the flavors of lime flesh and...   
3             US  Pineapple rind, lemon pith and orange blossom ...   
4             US  Much like the regular bottling from 2012, this...   
...          ...                                                ...   
129966   Germany  Notes of honeysuckle and cantaloupe sweeten th...   
129967        US  Citation is given as much as a decade of bottl...   
129968    France  Well-drained gravel soil gives this wine its c...   
129969    France  A dry style of Pinot Gris, this is crisp with ...   
129970    France  Big, rich and off-dry, this is powered by inte...   

                                   designation  points  price   
0                                 Vulkà Bianco      87    NaN  \
1                                     Avidagos      60   15.0   
2                                          NaN      87   14.0   
3                         Reserve Late Harvest      87   13.0   
4           Vintner's Reserve Wild Child Block      87   65.0   
...                                        ...     ...    ...   
129966  Brauneberger Juffer-Sonnenuhr Spätlese      90   28.0   
129967                                     NaN      90   75.0   
129968                                   Kritt      90   30.0   
129969                                     NaN      90   32.0   
129970           Lieu-dit Harth Cuvée Caroline      90   21.0   

                 province             region_1           region_2   
0       Sicily & Sardinia                 Etna                NaN  \
1                   Douro                  NaN                NaN   
2                  Oregon    Willamette Valley  Willamette Valley   
3                Michigan  Lake Michigan Shore                NaN   
4                  Oregon    Willamette Valley  Willamette Valley   
...                   ...                  ...                ...   
129966              Mosel                  NaN                NaN   
129967             Oregon               Oregon       Oregon Other   
129968             Alsace               Alsace                NaN   
129969             Alsace               Alsace                NaN   
129970             Alsace               Alsace                NaN   

               taster_name taster_twitter_handle   
0            Kerin O’Keefe          @kerinokeefe  \
1               Roger Voss            @vossroger   
2             Paul Gregutt           @paulgwine    
3       Alexander Peartree                   NaN   
4             Paul Gregutt           @paulgwine    
...                    ...                   ...   
129966  Anna Lee C. Iijima                   NaN   
129967        Paul Gregutt           @paulgwine    
129968          Roger Voss            @vossroger   
129969          Roger Voss            @vossroger   
129970          Roger Voss            @vossroger   

                                                    title         variety   
0                       Nicosia 2013 Vulkà Bianco  (Etna)     White Blend  \
1           Quinta dos Avidagos 2011 Avidagos Red (Douro)  Portuguese Red   
2           Rainstorm 2013 Pinot Gris (Willamette Valley)      Pinot Gris   
3       St. Julian 2013 Reserve Late Harvest Riesling ...        Riesling   
4       Sweet Cheeks 2012 Vintner's Reserve Wild Child...      Pinot Noir   
...                                                   ...             ...   
129966  Dr. H. Thanisch (Erben Müller-Burggraef) 2013 ...        Riesling   
129967                  Citation 2004 Pinot Noir (Oregon)      Pinot Noir   
129968  Domaine Gresser 2013 Kritt Gewurztraminer (Als...  Gewürztraminer   
129969      Domaine Marcel Deiss 2012 Pinot Gris (Alsace)      Pinot Gris   
129970  Domaine Schoffit 2012 Lieu-dit Harth Cuvée Car...  Gewürztraminer 

In Python, we can access the property of an object by accessing it as an attribute. A `book` object, for example, might have a `title` property, which we can access by calling `book.title`. Columns in a pandas DataFrame work in much the same way.

Hence to access the country property of reviews we can use:

In [6]:
reviews.points

0         87
1         60
2         87
3         87
4         87
          ..
129966    90
129967    90
129968    90
129969    90
129970    90
Name: points, Length: 129971, dtype: int64

If we have a Python dictionary, we can access its values using the indexing ([]) operator. We can do the same with columns in a DataFrame:

In [12]:
reviews['country']

0            Italy
1         Portugal
2               US
3               US
4               US
            ...   
129966     Germany
129967          US
129968      France
129969      France
129970      France
Name: country, Length: 129971, dtype: object

**Note :** These are the two ways of selecting a specific Series out of a DataFrame. Neither of them is more or less syntactically valid than the other, but the indexing operator [] does have the advantage that it can handle column names with reserved characters in them (e.g. if we had a country providence column, reviews.country providence wouldn't work).

In [13]:
reviews.country[4]

'US'

## Pandas - Indexing, Selecting, Assigning
### Indexing in Pandas

The indexing operator and attribute selection are nice because they work just like they do in the rest of the Python ecosystem. 

However, pandas has its own accessor operators, `loc` and `iloc`. For more advanced operations, these are the ones you're supposed to be using

- `loc` for label-based indexing
- `iloc` for Index-based indexing

#### Index-based selection

`iloc` selects data based on its numerical position in the data. iloc follows this paradigm.

To select the first row of data in a DataFrame, we may use the following:

In [18]:
reviews.iloc[0]

country                                                              Italy
description              Aromas include tropical fruit, broom, brimston...
designation                                                   Vulkà Bianco
points                                                                  87
price                                                                  NaN
                                               ...                        
taster_name                                                  Kerin O’Keefe
taster_twitter_handle                                         @kerinokeefe
title                                    Nicosia 2013 Vulkà Bianco  (Etna)
variety                                                        White Blend
winery                                                             Nicosia
Name: 0, Length: 13, dtype: object

**Note :** Both `loc` and `iloc` are row-first, column-second. This is the opposite of what we do in native Python, which is column-first, row-second.

This means that it's marginally easier to retrieve rows, and marginally harder to get retrieve columns. 

To get a column with iloc, we can do the following:

In [33]:
print(reviews.iloc[:, 0])

0            Italy
1         Portugal
2               US
3               US
4               US
            ...   
129966     Germany
129967          US
129968      France
129969      France
129970      France
Name: country, Length: 129971, dtype: object


On its own, the : operator, which also comes from native Python, means "everything". When combined with other selectors, however, it can be used to indicate a range of values. For example, to select the country column from just the first, second, and third row, we would do:

Or, to select just the second and third entries, we would do:

In [25]:
reviews.iloc[:3, :3]

country                                        description   designation
0     Italy  Aromas include tropical fruit, broom, brimston...  Vulkà Bianco
1  Portugal  This is ripe and fruity, a wine that is smooth...      Avidagos
2        US  Tart and snappy, the flavors of lime flesh and...           NaN

It's also possible to pass a list:

In [26]:
reviews.iloc[[0, 10, 20], 0]

0     Italy
10       US
20       US
Name: country, dtype: object

Finally, it's worth knowing that negative numbers can be used in selection.

In [27]:
reviews.iloc[-5:, 0]

129966    Germany
129967         US
129968     France
129969     France
129970     France
Name: country, dtype: object

#### Index-based selection

The second paradigm for attribute selection is the one followed by the `loc` operator: **label-based selection**. In this paradigm, it's the data index value, not its position, which matters.

For example, to get the first entry in `reviews`, we would now do the following:

In [31]:
reviews.loc[:, 'country']

0            Italy
1         Portugal
2               US
3               US
4               US
            ...   
129966     Germany
129967          US
129968      France
129969      France
129970      France
Name: country, Length: 129971, dtype: object

**Note :** iloc is conceptually simpler than loc because it ignores the dataset's indices. When we use iloc we treat the dataset like a big matrix (a list of lists), one that we have to index into by position. loc, by contrast, uses the information in the indices to do its work

Since your dataset usually has meaningful indices, it's usually easier to do things using loc instead. For example, here's one operation that's much easier using loc:

In [37]:
reviews.loc[:, ['taster_name', 'taster_twitter_handle', 'points']]
reviews.iloc[:, [8, 9, 3]]

taster_name taster_twitter_handle  points
0            Kerin O’Keefe          @kerinokeefe      87
1               Roger Voss            @vossroger      60
2             Paul Gregutt           @paulgwine       87
3       Alexander Peartree                   NaN      87
4             Paul Gregutt           @paulgwine       87
...                    ...                   ...     ...
129966  Anna Lee C. Iijima                   NaN      90
129967        Paul Gregutt           @paulgwine       90
129968          Roger Voss            @vossroger      90
129969          Roger Voss            @vossroger      90
129970          Roger Voss            @vossroger      90

[129971 rows x 3 columns]

#### Choosing between `loc` and `iloc`

When choosing or transitioning between `loc` and `iloc`, there is one "gotcha" worth keeping in mind, which is that the two methods use slightly different indexing schemes.

`iloc` uses the Python stdlib indexing scheme, where the first element of the range is included and the last one excluded. So 0:10 will select entries 0,...,9. `loc`, meanwhile, indexes inclusively. So 0:10 will select entries 0,...,10.

Why the change? Remember that `loc` can index any stdlib type: strings, for example. If we have a DataFrame with index values Apples, ..., Potatoes, ..., and we want to select "all the alphabetical fruit choices between Apples and Potatoes", then it's a heck of a lot more convenient to index `df.loc['Apples':'Potatoes']` than it is to index something like `df.loc['Apples', 'Potatoed']` (`t` coming after `s` in the alphabet).

This is particularly confusing when the DataFrame index is a simple numerical list, e.g. 0,...,1000. In this case df.iloc[0:1000] will return 1000 entries, while df.loc[0:1000] return 1001 of them! To get 1000 elements using loc, you will need to go one lower and ask for df.loc[0:999].

Otherwise, the semantics of using loc are the same as those for iloc.

#### Manipulating the index

Label-based selection derives its power from the labels in the index. Critically, the index we use is not immutable. We can manipulate the index in any way we see fit.

The `set_index()` method can be used to do the job. Here is what happens when we `set_index` to the `title` field:

In [38]:
reviews.set_index("title")

country   
title                                                          
Nicosia 2013 Vulkà Bianco  (Etna)                      Italy  \
Quinta dos Avidagos 2011 Avidagos Red (Douro)       Portugal   
Rainstorm 2013 Pinot Gris (Willamette Valley)             US   
St. Julian 2013 Reserve Late Harvest Riesling (...        US   
Sweet Cheeks 2012 Vintner's Reserve Wild Child ...        US   
...                                                      ...   
Dr. H. Thanisch (Erben Müller-Burggraef) 2013 B...   Germany   
Citation 2004 Pinot Noir (Oregon)                         US   
Domaine Gresser 2013 Kritt Gewurztraminer (Alsace)    France   
Domaine Marcel Deiss 2012 Pinot Gris (Alsace)         France   
Domaine Schoffit 2012 Lieu-dit Harth Cuvée Caro...    France   

                                                                                          description   
title                                                                                                   
Nicosia 2013 Vulkà Bianco  (Etna)                   Aromas include tropical fruit, broom, brimston...  \
Quinta dos Avidagos 2011 Avidagos Red (Douro)       This is ripe and fruity, a wine that is smooth...   
Rainstorm 2013 Pinot Gris (Willamette Valley)       Tart and snappy, the flavors of lime flesh and...   
St. Julian 2013 Reserve Late Harvest Riesling (...  Pineapple rind, lemon pith and orange blossom ...   
Sweet Cheeks 2012 Vintner's Reserve Wild Child ...  Much like the regular bottling from 2012, this...   
...                                                                                               ...   
Dr. H. Thanisch (Erben Müller-Burggraef) 2013 B...  Notes of honeysuckle and cantaloupe sweeten th...   
Citation 2004 Pinot Noir (Oregon)                   Citation is given as much as a decade of bottl...   
Domaine Gresser 2013 Kritt Gewurztraminer (Alsace)  Well-drained gravel soil gives this wine its c...   
Domaine Marcel Deiss 2012 Pinot Gris (Alsace)       A dry style of Pinot Gris, this is crisp with ...   
Domaine Schoffit 2012 Lieu-dit Harth Cuvée Caro...  Big, rich and off-dry, this is powered by inte...   

                                                                               designation   
title                                                                                        
Nicosia 2013 Vulkà Bianco  (Etna)                                             Vulkà Bianco  \
Quinta dos Avidagos 2011 Avidagos Red (Douro)                                     Avidagos   
Rainstorm 2013 Pinot Gris (Willamette Valley)                                          NaN   
St. Julian 2013 Reserve Late Harvest Riesling (...                    Reserve Late Harvest   
Sweet Cheeks 2012 Vintner's Reserve Wild Child ...      Vintner's Reserve Wild Child Block   
...                                                                                    ...   
Dr. H. Thanisch (Erben Müller-Burggraef) 2013 B...  Brauneberger Juffer-Sonnenuhr Spätlese   
Citation 2004 Pinot Noir (Oregon)                                                      NaN   
Domaine Gresser 2013 Kritt Gewurztraminer (Alsace)                                   Kritt   
Domaine Marcel Deiss 2012 Pinot Gris (Alsace)                                          NaN   
Domaine Schoffit 2012 Lieu-dit Harth Cuvée Caro...           Lieu-dit Harth Cuvée Caroline   

                                                    points  price   
title                                                               
Nicosia 2013 Vulkà Bianco  (Etna)                       87    NaN  \
Quinta dos Avidagos 2011 Avidagos Red (Douro)           60   15.0   
Rainstorm 2013 Pinot Gris (Willamette Valley)           87   14.0   
St. Julian 2013 Reserve Late Harvest Riesling (...      87   13.0   
Sweet Cheeks 2012 Vintner's Reserve Wild Child ...      87   65.0   
...                                                    ...    ...   
Dr. H. Thanisch (Erben Müller-Burggraef) 2013 B...      90   28.0   
Citation 2

#### Conditional selection

So far we've been indexing various strides of data, using structural properties of the DataFrame itself. To do *interesting* things with the data, however, we often need to ask questions based on conditions.

For example, suppose that we're interested specifically in better-than-average wines produced in Italy.

We can start by checking if each wine is Italian or not:

In [39]:
reviews.country == 'Italy'

0          True
1         False
2         False
3         False
4         False
          ...  
129966    False
129967    False
129968    False
129969    False
129970    False
Name: country, Length: 129971, dtype: bool

This operation produced a Series of True/False booleans based on the country of each record. This result can then be used inside of loc to select the relevant data:

In [40]:
italian_wines = reviews.loc[reviews.country == 'Italy']
italian_wines

country                                        description   
0        Italy  Aromas include tropical fruit, broom, brimston...  \
6        Italy  Here's a bright, informal red that opens with ...   
13       Italy  This is dominated by oak and oak-driven aromas...   
22       Italy  Delicate aromas recall white flower and citrus...   
24       Italy  Aromas of prune, blackcurrant, toast and oak c...   
...        ...                                                ...   
129929   Italy  This luminous sparkler has a sweet, fruit-forw...   
129943   Italy  A blend of Nero d'Avola and Syrah, this convey...   
129947   Italy  A blend of 65% Cabernet Sauvignon, 30% Merlot ...   
129961   Italy  Intense aromas of wild cherry, baking spice, t...   
129962   Italy  Blackberry, cassis, grilled herb and toasted a...   

                      designation  points  price           province   
0                    Vulkà Bianco      87    NaN  Sicily & Sardinia  \
6                         Belsito      87   16.0  Sicily & Sardinia   
13                          Rosso      87    NaN  Sicily & Sardinia   
22                      Ficiligno      87   19.0  Sicily & Sardinia   
24                          Aynat      87   35.0  Sicily & Sardinia   
...                           ...     ...    ...                ...   
129929                        NaN      91   38.0             Veneto   
129943                    Adènzia      90   29.0  Sicily & Sardinia   
129947                   Symposio      90   20.0  Sicily & Sardinia   
129961                        NaN      90   30.0  Sicily & Sardinia   
129962  Sàgana Tenuta San Giacomo      90   40.0  Sicily & Sardinia   

                              region_1 region_2    taster_name   
0                                 Etna      NaN  Kerin O’Keefe  \
6                             Vittoria      NaN  Kerin O’Keefe   
13                                Etna      NaN  Kerin O’Keefe   
22                             Sicilia      NaN  Kerin O’Keefe   
24                             Sicilia      NaN  Kerin O’Keefe   
...                                ...      ...            ...   
129929  Prosecco Superiore di Cartizze      NaN            NaN   
129943                         Sicilia      NaN  Kerin O’Keefe   
129947                 Terre Siciliane      NaN  Kerin O’Keefe   
129961                         Sicilia      NaN  Kerin O’Keefe   
129962                         Sicilia      NaN  Kerin O’Keefe   

       taster_twitter_handle   
0               @kerinokeefe  \
6               @kerinokeefe   
13              @kerinokeefe   
22              @kerinokeefe   
24              @kerinokeefe   
...                      ...   
129929                   NaN   
129943          @kerinokeefe   
129947          @kerinokeefe   
129961          @kerinokeefe   
129962          @kerinokeefe   

                                                    title            variety   
0                       Nicosia 2013 Vulkà Bianco  (Etna)        White Blend  \
6        Terre di Giurfo 2013 Belsito Frappato (Vittoria)           Frappato   
13                 Masseria Setteporte 2012 Rosso  (Etna)  Nerello Mascalese   
22      Baglio di Pianetto 2007 Ficiligno White (Sicilia)        White Blend   
24            Canicattì 2009 Aynat Nero d'Avola (Sicilia)       Nero d'Avola   
...                                                   ...                ...   
129929  Col Vetoraz Spumanti NV  Prosecco Superiore di...           Prosecco   
129943  Baglio del Cristo di Campobello 2012 Adènzia R...          Red Blend   
129947  Feudo Principi di Butera 2012 Symposio Red (Te...          Red Blend   
129961                        COS 2013 Frappato (Sicilia)           Frappato   
129962  Cusumano 2012 Sàgana Tenuta San Giacomo Nero d...       Nero d'Avola   

                                 winery  
0                               Nicosia  
6                       Terre di Giurfo  
13                  Masseria Setteporte  
22                

**Note :** This DataFrame has ~20,000 rows. The original had ~130,000. That means that around 15% of wines originate from Italy.

We also wanted to know which ones are better than average. Wines are reviewed on a 80-to-100 point scale, so this could mean wines that accrued at least 90 points.

We can use the ampersand (`&`) to bring the two questions together:

In [41]:
reviews.loc[(reviews.country == 'Italy') & (reviews.points >= 90)]

country                                        description   
120      Italy  Slightly backward, particularly given the vint...  \
130      Italy  At the first it was quite muted and subdued, b...   
133      Italy  Einaudi's wines have been improving lately, an...   
135      Italy  The color is just beginning to show signs of b...   
140      Italy  A big, fat, luscious wine with plenty of toast...   
...        ...                                                ...   
129929   Italy  This luminous sparkler has a sweet, fruit-forw...   
129943   Italy  A blend of Nero d'Avola and Syrah, this convey...   
129947   Italy  A blend of 65% Cabernet Sauvignon, 30% Merlot ...   
129961   Italy  Intense aromas of wild cherry, baking spice, t...   
129962   Italy  Blackberry, cassis, grilled herb and toasted a...   

                      designation  points  price           province   
120           Bricco Rocche Prapó      92   70.0           Piedmont  \
130         Bricco Rocche Brunate      91   70.0           Piedmont   
133                           NaN      91   68.0           Piedmont   
135                        Sorano      91   60.0           Piedmont   
140                   Costa Bruna      90   26.0           Piedmont   
...                           ...     ...    ...                ...   
129929                        NaN      91   38.0             Veneto   
129943                    Adènzia      90   29.0  Sicily & Sardinia   
129947                   Symposio      90   20.0  Sicily & Sardinia   
129961                        NaN      90   30.0  Sicily & Sardinia   
129962  Sàgana Tenuta San Giacomo      90   40.0  Sicily & Sardinia   

                              region_1 region_2    taster_name   
120                             Barolo      NaN            NaN  \
130                             Barolo      NaN            NaN   
133                             Barolo      NaN            NaN   
135                             Barolo      NaN            NaN   
140                     Barbera d'Alba      NaN            NaN   
...                                ...      ...            ...   
129929  Prosecco Superiore di Cartizze      NaN            NaN   
129943                         Sicilia      NaN  Kerin O’Keefe   
129947                 Terre Siciliane      NaN  Kerin O’Keefe   
129961                         Sicilia      NaN  Kerin O’Keefe   
129962                         Sicilia      NaN  Kerin O’Keefe   

       taster_twitter_handle   
120                      NaN  \
130                      NaN   
133                      NaN   
135                      NaN   
140                      NaN   
...                      ...   
129929                   NaN   
129943          @kerinokeefe   
129947          @kerinokeefe   
129961          @kerinokeefe   
129962          @kerinokeefe   

                                                    title       variety   
120            Ceretto 2003 Bricco Rocche Prapó  (Barolo)      Nebbiolo  \
130          Ceretto 2003 Bricco Rocche Brunate  (Barolo)      Nebbiolo   
133                     Poderi Luigi Einaudi 2003  Barolo      Nebbiolo   
135                 Giacomo Ascheri 2001 Sorano  (Barolo)      Nebbiolo   
140       Poderi Colla 2005 Costa Bruna  (Barbera d'Alba)       Barbera   
...                                                   ...           ...   
129929  Col Vetoraz Spumanti NV  Prosecco Superiore di...      Prosecco   
129943  Baglio del Cristo di Campobello 2012 Adènzia R...     Red Blend   
129947  Feudo Principi di Butera 2012 Symposio Red (Te...     Red Blend   
129961                        COS 2013 Frappato (Sicilia)      Frappato   
129962  Cusumano 2012 Sàgana Tenuta San Giacomo Nero d...  Nero d'Avola   

                                 winery  
120                             Ceretto  
130                             Ceretto  
133                Poderi Luigi Einaudi  
135                     Giacomo Ascheri  
140                        Poderi Co

Suppose we'll buy any wine that's made in Italy or which is rated above average. For this we use a pipe (|):

In [44]:
reviews.loc[(reviews.country == 'Italy') | (reviews.points >= 90)]

country                                        description   
0         Italy  Aromas include tropical fruit, broom, brimston...  \
6         Italy  Here's a bright, informal red that opens with ...   
13        Italy  This is dominated by oak and oak-driven aromas...   
22        Italy  Delicate aromas recall white flower and citrus...   
24        Italy  Aromas of prune, blackcurrant, toast and oak c...   
...         ...                                                ...   
129966  Germany  Notes of honeysuckle and cantaloupe sweeten th...   
129967       US  Citation is given as much as a decade of bottl...   
129968   France  Well-drained gravel soil gives this wine its c...   
129969   France  A dry style of Pinot Gris, this is crisp with ...   
129970   France  Big, rich and off-dry, this is powered by inte...   

                                   designation  points  price   
0                                 Vulkà Bianco      87    NaN  \
6                                      Belsito      87   16.0   
13                                       Rosso      87    NaN   
22                                   Ficiligno      87   19.0   
24                                       Aynat      87   35.0   
...                                        ...     ...    ...   
129966  Brauneberger Juffer-Sonnenuhr Spätlese      90   28.0   
129967                                     NaN      90   75.0   
129968                                   Kritt      90   30.0   
129969                                     NaN      90   32.0   
129970           Lieu-dit Harth Cuvée Caroline      90   21.0   

                 province  region_1      region_2         taster_name   
0       Sicily & Sardinia      Etna           NaN       Kerin O’Keefe  \
6       Sicily & Sardinia  Vittoria           NaN       Kerin O’Keefe   
13      Sicily & Sardinia      Etna           NaN       Kerin O’Keefe   
22      Sicily & Sardinia   Sicilia           NaN       Kerin O’Keefe   
24      Sicily & Sardinia   Sicilia           NaN       Kerin O’Keefe   
...                   ...       ...           ...                 ...   
129966              Mosel       NaN           NaN  Anna Lee C. Iijima   
129967             Oregon    Oregon  Oregon Other        Paul Gregutt   
129968             Alsace    Alsace           NaN          Roger Voss   
129969             Alsace    Alsace           NaN          Roger Voss   
129970             Alsace    Alsace           NaN          Roger Voss   

       taster_twitter_handle   
0               @kerinokeefe  \
6               @kerinokeefe   
13              @kerinokeefe   
22              @kerinokeefe   
24              @kerinokeefe   
...                      ...   
129966                   NaN   
129967           @paulgwine    
129968            @vossroger   
129969            @vossroger   
129970            @vossroger   

                                                    title            variety   
0                       Nicosia 2013 Vulkà Bianco  (Etna)        White Blend  \
6        Terre di Giurfo 2013 Belsito Frappato (Vittoria)           Frappato   
13                 Masseria Setteporte 2012 Rosso  (Etna)  Nerello Mascalese   
22      Baglio di Pianetto 2007 Ficiligno White (Sicilia)        White Blend   
24            Canicattì 2009 Aynat Nero d'Avola (Sicilia)       Nero d'Avola   
...                                                   ...                ...   
129966  Dr. H. Thanisch (Erben Müller-Burggraef) 2013 ...           Riesling   
129967                  Citation 2004 Pinot Noir (Oregon)         Pinot Noir   
129968  Domaine Gresser 2013 Kritt Gewurztraminer (Als...     Gewürztraminer   
129969      Domaine Marcel Deiss 2012 Pinot Gris (Alsace)         Pinot Gris   
129970  Domaine Schoffit 2012 Lieu-dit Harth Cuvée Car...     Gewürztraminer   

                                          winery  
0                                        Nicosia  
6                                Terre di Giurfo  
13        

## Pandas Built-in conditinal selectors

| Selector | Description |
|----------|-------------|
| `isin()` | Selects rows where the column contains one or more values |
| `isnull()` | Selects rows where the column is null |
| `notnull()` | Selects rows where the column is not null |
| `between()` | Selects rows where the column is between two values |
| `any()` | Selects rows where the column contains at least one non-null value |
| `all()` | Selects rows where the column contains all non-null values |

In [45]:
reviews.loc[reviews.country.isin(['Italy', 'France'])]

country                                        description   
0        Italy  Aromas include tropical fruit, broom, brimston...  \
6        Italy  Here's a bright, informal red that opens with ...   
7       France  This dry and restrained wine offers spice in p...   
9       France  This has great depth of flavor with its fresh ...   
11      France  This is a dry wine, very spicy, with a tight, ...   
...        ...                                                ...   
129964  France  Initially quite muted, this wine slowly develo...   
129965  France  While it's rich, this beautiful dry wine also ...   
129968  France  Well-drained gravel soil gives this wine its c...   
129969  France  A dry style of Pinot Gris, this is crisp with ...   
129970  France  Big, rich and off-dry, this is powered by inte...   

                          designation  points  price           province   
0                        Vulkà Bianco      87    NaN  Sicily & Sardinia  \
6                             Belsito      87   16.0  Sicily & Sardinia   
7                                 NaN      87   24.0             Alsace   
9                         Les Natures      87   27.0             Alsace   
11                                NaN      87   30.0             Alsace   
...                               ...     ...    ...                ...   
129964   Domaine Saint-Rémy Herrenweg      90    NaN             Alsace   
129965    Seppi Landmann Vallée Noble      90   28.0             Alsace   
129968                          Kritt      90   30.0             Alsace   
129969                            NaN      90   32.0             Alsace   
129970  Lieu-dit Harth Cuvée Caroline      90   21.0             Alsace   

        region_1 region_2    taster_name taster_twitter_handle   
0           Etna      NaN  Kerin O’Keefe          @kerinokeefe  \
6       Vittoria      NaN  Kerin O’Keefe          @kerinokeefe   
7         Alsace      NaN     Roger Voss            @vossroger   
9         Alsace      NaN     Roger Voss            @vossroger   
11        Alsace      NaN     Roger Voss            @vossroger   
...          ...      ...            ...                   ...   
129964    Alsace      NaN     Roger Voss            @vossroger   
129965    Alsace      NaN     Roger Voss            @vossroger   
129968    Alsace      NaN     Roger Voss            @vossroger   
129969    Alsace      NaN     Roger Voss            @vossroger   
129970    Alsace      NaN     Roger Voss            @vossroger   

                                                    title         variety   
0                       Nicosia 2013 Vulkà Bianco  (Etna)     White Blend  \
6        Terre di Giurfo 2013 Belsito Frappato (Vittoria)        Frappato   
7                   Trimbach 2012 Gewurztraminer (Alsace)  Gewürztraminer   
9       Jean-Baptiste Adam 2012 Les Natures Pinot Gris...      Pinot Gris   
11                Leon Beyer 2012 Gewurztraminer (Alsace)  Gewürztraminer   
...                                                   ...             ...   
129964  Domaine Ehrhart 2013 Domaine Saint-Rémy Herren...  Gewürztraminer   
129965  Domaine Rieflé-Landmann 2013 Seppi Landmann Va...      Pinot Gris   
129968  Domaine Gresser 2013 Kritt Gewurztraminer (Als...  Gewürztraminer   
129969      Domaine Marcel Deiss 2012 Pinot Gris (Alsace)      Pinot Gris   
129970  Domaine Schoffit 2012 Lieu-dit Harth Cuvée Car...  Gewürztraminer   

                         winery  
0                       Nicosia  
6               Terre di Giurfo  
7                      Trimbach  
9            Jean-Baptiste Adam  
11                   Leon Beyer  
...                         ...  
129964          Domaine Ehrhart  
129965  Domaine Rieflé-Landmann  
129968          Domaine Gresser  
129969     Domaine Marcel Deiss  
129970         Domaine Schoffit  

[41633 rows x 13 columns]

**Note :** The second is isnull (and its companion notnull). These methods let you highlight values which are (or are not) empty (NaN). For example, to filter out wines lacking a price tag in the dataset, here's what we would do:

In [48]:
reviews.loc[reviews.designation.notnull()]

country                                        description   
0          Italy  Aromas include tropical fruit, broom, brimston...  \
1       Portugal  This is ripe and fruity, a wine that is smooth...   
3             US  Pineapple rind, lemon pith and orange blossom ...   
4             US  Much like the regular bottling from 2012, this...   
5          Spain  Blackberry and raspberry aromas show a typical...   
...          ...                                                ...   
129964    France  Initially quite muted, this wine slowly develo...   
129965    France  While it's rich, this beautiful dry wine also ...   
129966   Germany  Notes of honeysuckle and cantaloupe sweeten th...   
129968    France  Well-drained gravel soil gives this wine its c...   
129970    France  Big, rich and off-dry, this is powered by inte...   

                                   designation  points  price   
0                                 Vulkà Bianco      87    NaN  \
1                                     Avidagos      60   15.0   
3                         Reserve Late Harvest      87   13.0   
4           Vintner's Reserve Wild Child Block      87   65.0   
5                                 Ars In Vitro      87   15.0   
...                                        ...     ...    ...   
129964            Domaine Saint-Rémy Herrenweg      90    NaN   
129965             Seppi Landmann Vallée Noble      90   28.0   
129966  Brauneberger Juffer-Sonnenuhr Spätlese      90   28.0   
129968                                   Kritt      90   30.0   
129970           Lieu-dit Harth Cuvée Caroline      90   21.0   

                 province             region_1           region_2   
0       Sicily & Sardinia                 Etna                NaN  \
1                   Douro                  NaN                NaN   
3                Michigan  Lake Michigan Shore                NaN   
4                  Oregon    Willamette Valley  Willamette Valley   
5          Northern Spain              Navarra                NaN   
...                   ...                  ...                ...   
129964             Alsace               Alsace                NaN   
129965             Alsace               Alsace                NaN   
129966              Mosel                  NaN                NaN   
129968             Alsace               Alsace                NaN   
129970             Alsace               Alsace                NaN   

               taster_name taster_twitter_handle   
0            Kerin O’Keefe          @kerinokeefe  \
1               Roger Voss            @vossroger   
3       Alexander Peartree                   NaN   
4             Paul Gregutt           @paulgwine    
5        Michael Schachner           @wineschach   
...                    ...                   ...   
129964          Roger Voss            @vossroger   
129965          Roger Voss            @vossroger   
129966  Anna Lee C. Iijima                   NaN   
129968          Roger Voss            @vossroger   
129970          Roger Voss            @vossroger   

                                                    title             variety   
0                       Nicosia 2013 Vulkà Bianco  (Etna)         White Blend  \
1           Quinta dos Avidagos 2011 Avidagos Red (Douro)      Portuguese Red   
3       St. Julian 2013 Reserve Late Harvest Riesling ...            Riesling   
4       Sweet Cheeks 2012 Vintner's Reserve Wild Child...          Pinot Noir   
5       Tandem 2011 Ars In Vitro Tempranillo-Merlot (N...  Tempranillo-Merlot   
...                                                   ...                 ...   
129964  Domaine Ehrhart 2013 Domaine Saint-Rémy Herren...      Gewürztraminer   
129965  Domaine Rieflé-Landmann 2013 Seppi Landmann Va...          Pinot Gris   
129966  Dr. H. Thanisch (Erben Müller-Burggraef) 2013 ...            Riesling   
129968  Domaine Gresser 2013 Kritt Gewurztraminer (Als...      Gewürztraminer   
129970  Domaine Schoffit 2012 

#### Assagning data

Going the other way, assigning data to a DataFrame is easy. You can assign either a constant value:

In [49]:
reviews['critic'] = 'everyone'
reviews

country                                        description   
0          Italy  Aromas include tropical fruit, broom, brimston...  \
1       Portugal  This is ripe and fruity, a wine that is smooth...   
2             US  Tart and snappy, the flavors of lime flesh and...   
3             US  Pineapple rind, lemon pith and orange blossom ...   
4             US  Much like the regular bottling from 2012, this...   
...          ...                                                ...   
129966   Germany  Notes of honeysuckle and cantaloupe sweeten th...   
129967        US  Citation is given as much as a decade of bottl...   
129968    France  Well-drained gravel soil gives this wine its c...   
129969    France  A dry style of Pinot Gris, this is crisp with ...   
129970    France  Big, rich and off-dry, this is powered by inte...   

                                   designation  points  price   
0                                 Vulkà Bianco      87    NaN  \
1                                     Avidagos      60   15.0   
2                                          NaN      87   14.0   
3                         Reserve Late Harvest      87   13.0   
4           Vintner's Reserve Wild Child Block      87   65.0   
...                                        ...     ...    ...   
129966  Brauneberger Juffer-Sonnenuhr Spätlese      90   28.0   
129967                                     NaN      90   75.0   
129968                                   Kritt      90   30.0   
129969                                     NaN      90   32.0   
129970           Lieu-dit Harth Cuvée Caroline      90   21.0   

                 province             region_1           region_2   
0       Sicily & Sardinia                 Etna                NaN  \
1                   Douro                  NaN                NaN   
2                  Oregon    Willamette Valley  Willamette Valley   
3                Michigan  Lake Michigan Shore                NaN   
4                  Oregon    Willamette Valley  Willamette Valley   
...                   ...                  ...                ...   
129966              Mosel                  NaN                NaN   
129967             Oregon               Oregon       Oregon Other   
129968             Alsace               Alsace                NaN   
129969             Alsace               Alsace                NaN   
129970             Alsace               Alsace                NaN   

               taster_name taster_twitter_handle   
0            Kerin O’Keefe          @kerinokeefe  \
1               Roger Voss            @vossroger   
2             Paul Gregutt           @paulgwine    
3       Alexander Peartree                   NaN   
4             Paul Gregutt           @paulgwine    
...                    ...                   ...   
129966  Anna Lee C. Iijima                   NaN   
129967        Paul Gregutt           @paulgwine    
129968          Roger Voss            @vossroger   
129969          Roger Voss            @vossroger   
129970          Roger Voss            @vossroger   

                                                    title         variety   
0                       Nicosia 2013 Vulkà Bianco  (Etna)     White Blend  \
1           Quinta dos Avidagos 2011 Avidagos Red (Douro)  Portuguese Red   
2           Rainstorm 2013 Pinot Gris (Willamette Valley)      Pinot Gris   
3       St. Julian 2013 Reserve Late Harvest Riesling ...        Riesling   
4       Sweet Cheeks 2012 Vintner's Reserve Wild Child...      Pinot Noir   
...                                                   ...             ...   
129966  Dr. H. Thanisch (Erben Müller-Burggraef) 2013 ...        Riesling   
129967                  Citation 2004 Pinot Noir (Oregon)      Pinot Noir   
129968  Domaine Gresser 2013 Kritt Gewurztraminer (Als...  Gewürztraminer   
129969      Domaine Marcel Deiss 2012 Pinot Gris (Alsace)      Pinot Gris   
129970  Domaine Schoffit 2012 Lieu-dit Harth Cuvée Car...  Gewürztraminer 

In [57]:
# Or with iterable values
back_index = list(range(len(reviews)-1, -1, -1))
print(len(back_index))
reviews['country_italy'] = reviews.country == "Italy"
reviews['country_italy']
reviews

129971


country                                        description   
0          Italy  Aromas include tropical fruit, broom, brimston...  \
1       Portugal  This is ripe and fruity, a wine that is smooth...   
2             US  Tart and snappy, the flavors of lime flesh and...   
3             US  Pineapple rind, lemon pith and orange blossom ...   
4             US  Much like the regular bottling from 2012, this...   
...          ...                                                ...   
129966   Germany  Notes of honeysuckle and cantaloupe sweeten th...   
129967        US  Citation is given as much as a decade of bottl...   
129968    France  Well-drained gravel soil gives this wine its c...   
129969    France  A dry style of Pinot Gris, this is crisp with ...   
129970    France  Big, rich and off-dry, this is powered by inte...   

                                   designation  points  price   
0                                 Vulkà Bianco      87    NaN  \
1                                     Avidagos      60   15.0   
2                                          NaN      87   14.0   
3                         Reserve Late Harvest      87   13.0   
4           Vintner's Reserve Wild Child Block      87   65.0   
...                                        ...     ...    ...   
129966  Brauneberger Juffer-Sonnenuhr Spätlese      90   28.0   
129967                                     NaN      90   75.0   
129968                                   Kritt      90   30.0   
129969                                     NaN      90   32.0   
129970           Lieu-dit Harth Cuvée Caroline      90   21.0   

                 province             region_1           region_2   
0       Sicily & Sardinia                 Etna                NaN  \
1                   Douro                  NaN                NaN   
2                  Oregon    Willamette Valley  Willamette Valley   
3                Michigan  Lake Michigan Shore                NaN   
4                  Oregon    Willamette Valley  Willamette Valley   
...                   ...                  ...                ...   
129966              Mosel                  NaN                NaN   
129967             Oregon               Oregon       Oregon Other   
129968             Alsace               Alsace                NaN   
129969             Alsace               Alsace                NaN   
129970             Alsace               Alsace                NaN   

               taster_name taster_twitter_handle   
0            Kerin O’Keefe          @kerinokeefe  \
1               Roger Voss            @vossroger   
2             Paul Gregutt           @paulgwine    
3       Alexander Peartree                   NaN   
4             Paul Gregutt           @paulgwine    
...                    ...                   ...   
129966  Anna Lee C. Iijima                   NaN   
129967        Paul Gregutt           @paulgwine    
129968          Roger Voss            @vossroger   
129969          Roger Voss            @vossroger   
129970          Roger Voss            @vossroger   

                                                    title         variety   
0                       Nicosia 2013 Vulkà Bianco  (Etna)     White Blend  \
1           Quinta dos Avidagos 2011 Avidagos Red (Douro)  Portuguese Red   
2           Rainstorm 2013 Pinot Gris (Willamette Valley)      Pinot Gris   
3       St. Julian 2013 Reserve Late Harvest Riesling ...        Riesling   
4       Sweet Cheeks 2012 Vintner's Reserve Wild Child...      Pinot Noir   
...                                                   ...             ...   
129966  Dr. H. Thanisch (Erben Müller-Burggraef) 2013 ...        Riesling   
129967                  Citation 2004 Pinot Noir (Oregon)      Pinot Noir   
129968  Domaine Gresser 2013 Kritt Gewurztraminer (Als...  Gewürztraminer   
129969      Domaine Marcel Deiss 2012 Pinot Gris (Alsace)      Pinot Gris   
129970  Domaine Schoffit 2012 Lieu-dit Harth Cuvée Car...  Gewürztraminer 

## Summary functions and Maps
### Introduction

In the last tutorial, we learned how to select relevant data out of a DataFrame or Series. Plucking the right data out of our data representation is critical to getting work done, as we demonstrated in the exercises.

However, the data does not always come out of memory in the format we want it in right out of the bat. Sometimes we have to do some more work ourselves to reformat it for the task at hand. This tutorial will cover different operations we can apply to our data to get the input "just right".

In [3]:
import numpy as np
import pandas as pd

### Summary functions

Pandas provides many simple "summary functions" (not an official name) which restructure the data in some useful way. 

For example, consider the `describe()` method:

In [4]:
pd.set_option('display.max_rows', 10)
reviews = pd.read_csv("./data/numpy_pandas/dataset/wine-reviews/winemag-data-130k-v2.csv", index_col=0)
reviews.describe(include=[np.number])

points          price
count  129971.000000  120975.000000
mean       88.446930      35.363389
std         3.040752      41.022218
min        60.000000       4.000000
25%        86.000000      17.000000
50%        88.000000      25.000000
75%        91.000000      42.000000
max       100.000000    3300.000000

This method generates a high-level summary of the attributes of the given column. It is type-aware, meaning that its output changes based on the data type of the input. The output above only makes sense for numerical data; for string data here's what we get:

In [25]:
reviews.taster_name.describe()

count         103727
unique            19
top       Roger Voss
freq           25514
Name: taster_name, dtype: object

If you want to get some particular simple summary statistic about a column in a DataFrame or a Series, there is usually a helpful pandas function that makes it happen.

For example, to see the mean of the points allotted (e.g. how well an averagely rated wine does), we can use the mean() function:

In [31]:
reviews.points.mean()

88.44693046910464

To see a list of unique values we can use the unique() function:

In [32]:
reviews.points

0         87
1         60
2         87
3         87
4         87
          ..
129966    90
129967    90
129968    90
129969    90
129970    90
Name: points, Length: 129971, dtype: int64

To see a list of unique values and how often they occur in the dataset, we can use the value_counts() method:

In [28]:
reviews.taster_name.value_counts()

taster_name
Roger Voss            25514
Michael Schachner     15134
Kerin O’Keefe         10776
Virginie Boone         9537
Paul Gregutt           9532
                      ...  
Jeff Jenssen            491
Alexander Peartree      415
Carrie Dykes            139
Fiona Adams              27
Christina Pickard         6
Name: count, Length: 19, dtype: int64

### Maps

A map is a term, borrowed from mathematics, for a function that takes one set of values and "maps" them to another set of values. In data science we often have a need for creating new representations from existing data, or for transforming data from the format it is in now to the format that we want it to be in later. Maps are what handle this work, making them extremely important for getting your work done!

There are two mapping methods that you will use often.
  - `map()`
  - `apply()`

`map()`  is the first, and slightly simpler one. For example, suppose that we wanted to remean the scores the wines received to 0. We can do this as follows:

In [8]:
review_points_mean = reviews.points.mean()
reviews.points.map(lambda p: p - review_points_mean)

0         -1.44693
1        -28.44693
2         -1.44693
3         -1.44693
4         -1.44693
            ...   
129966     1.55307
129967     1.55307
129968     1.55307
129969     1.55307
129970     1.55307
Name: points, Length: 129971, dtype: float64

**Note :** The function you pass to map() should expect a single value from the Series (a point value, in the above example), and return a transformed version of that value. 

map() returns a new Series where all the values have been transformed by your function

apply() is the equivalent method if we want to transform a whole DataFrame by calling a custom method on each row.

In [15]:
def remean_points(row):
    row.points = row.points - review_points_mean
    return row

reviews2 = reviews.apply(remean_points, axis='columns)
reviews2

country                                        description   
0          Italy  Aromas include tropical fruit, broom, brimston...  \
1       Portugal  This is ripe and fruity, a wine that is smooth...   
2             US  Tart and snappy, the flavors of lime flesh and...   
3             US  Pineapple rind, lemon pith and orange blossom ...   
4             US  Much like the regular bottling from 2012, this...   
...          ...                                                ...   
129966   Germany  Notes of honeysuckle and cantaloupe sweeten th...   
129967        US  Citation is given as much as a decade of bottl...   
129968    France  Well-drained gravel soil gives this wine its c...   
129969    France  A dry style of Pinot Gris, this is crisp with ...   
129970    France  Big, rich and off-dry, this is powered by inte...   

                                   designation    points  price   
0                                 Vulkà Bianco  -1.44693    NaN  \
1                                     Avidagos -28.44693   15.0   
2                                          NaN  -1.44693   14.0   
3                         Reserve Late Harvest  -1.44693   13.0   
4           Vintner's Reserve Wild Child Block  -1.44693   65.0   
...                                        ...       ...    ...   
129966  Brauneberger Juffer-Sonnenuhr Spätlese   1.55307   28.0   
129967                                     NaN   1.55307   75.0   
129968                                   Kritt   1.55307   30.0   
129969                                     NaN   1.55307   32.0   
129970           Lieu-dit Harth Cuvée Caroline   1.55307   21.0   

                 province             region_1           region_2   
0       Sicily & Sardinia                 Etna                NaN  \
1                   Douro                  NaN                NaN   
2                  Oregon    Willamette Valley  Willamette Valley   
3                Michigan  Lake Michigan Shore                NaN   
4                  Oregon    Willamette Valley  Willamette Valley   
...                   ...                  ...                ...   
129966              Mosel                  NaN                NaN   
129967             Oregon               Oregon       Oregon Other   
129968             Alsace               Alsace                NaN   
129969             Alsace               Alsace                NaN   
129970             Alsace               Alsace                NaN   

               taster_name taster_twitter_handle   
0            Kerin O’Keefe          @kerinokeefe  \
1               Roger Voss            @vossroger   
2             Paul Gregutt           @paulgwine    
3       Alexander Peartree                   NaN   
4             Paul Gregutt           @paulgwine    
...                    ...                   ...   
129966  Anna Lee C. Iijima                   NaN   
129967        Paul Gregutt           @paulgwine    
129968          Roger Voss            @vossroger   
129969          Roger Voss            @vossroger   
129970          Roger Voss            @vossroger   

                                                    title         variety   
0                       Nicosia 2013 Vulkà Bianco  (Etna)     White Blend  \
1           Quinta dos Avidagos 2011 Avidagos Red (Douro)  Portuguese Red   
2           Rainstorm 2013 Pinot Gris (Willamette Valley)      Pinot Gris   
3       St. Julian 2013 Reserve Late Harvest Riesling ...        Riesling   
4       Sweet Cheeks 2012 Vintner's Reserve Wild Child...      Pinot Noir   
...                                                   ...             ...   
129966  Dr. H. Thanisch (Erben Müller-Burggraef) 2013 ...        Riesling   
129967                  Citation 2004 Pinot Noir (Oregon)      Pinot Noir   
129968  Domaine Gresser 2013 Kritt Gewurztraminer (Als...  Gewürztraminer   
129969      Domaine Marcel Deiss 2012 Pinot Gris (Alsace)      Pinot Gris   
129970  Domaine Schoffit 2012 Lieu-dit Harth Cuvée

If we had called reviews.apply() with axis='index', then instead of passing a function to transform each row, we would need to give a function to transform each column.

**Note :** that map() and apply() return new, transformed Series and DataFrames, respectively. They don't modify the original data they're called on. If we look at the first row of reviews, we can see that it still has its original points value.

In [12]:
reviews.head(1)

country                                        description   designation   
0   Italy  Aromas include tropical fruit, broom, brimston...  Vulkà Bianco  \

   points  price           province region_1 region_2    taster_name   
0      87    NaN  Sicily & Sardinia     Etna      NaN  Kerin O’Keefe  \

  taster_twitter_handle                              title      variety   
0          @kerinokeefe  Nicosia 2013 Vulkà Bianco  (Etna)  White Blend  \

    winery  
0  Nicosia

Pandas provides many common mapping operations as built-ins. For example, here's a faster way of remeaning our points column:

In [16]:
review_points_mean = reviews.points.mean()
reviews.points - review_points_mean

0         -1.44693
1        -28.44693
2         -1.44693
3         -1.44693
4         -1.44693
            ...   
129966     1.55307
129967     1.55307
129968     1.55307
129969     1.55307
129970     1.55307
Name: points, Length: 129971, dtype: float64

In this code we are performing an operation between a lot of values on the left-hand side (everything in the Series) and a single value on the right-hand side (the mean value). Pandas looks at this expression and figures out that we must mean to subtract that mean value from every value in the dataset.

Pandas will also understand what to do if we perform these operations between Series of equal length. For example, an easy way of combining country and region information in the dataset would be to do the following:

In [17]:
reviews.country + " - " + reviews.region_1

0                     Italy - Etna
1                              NaN
2           US - Willamette Valley
3         US - Lake Michigan Shore
4           US - Willamette Valley
                    ...           
129966                         NaN
129967                 US - Oregon
129968             France - Alsace
129969             France - Alsace
129970             France - Alsace
Length: 129971, dtype: object

These operators are faster than map() or apply() because they use speed ups built into pandas. All of the standard Python operators (>, <, ==, and so on) work in this manner.

However, they are not as flexible as map() or apply(), which can do more advanced things, like applying conditional logic, which cannot be done with addition and subtraction alone.

## Pandas - Grouping and sorting
### Introduction

Maps allow us to transform data in a DataFrame or Series one value at a time for an entire column. However, often we want to group our data, and then do something specific to the group the data is in.

As you'll learn, we do this with the groupby() operation. We'll also cover some additional topics, such as more complex ways to index your DataFrames, along with how to sort your data.

### Groupwise analysis

One function we've been using heavily thus far is the value_counts() function. We can replicate what value_counts() does by doing the following:

In [25]:
reviews.groupby('points').points.count()

points
60        1
80      397
81      692
82     1836
83     3025
       ... 
96      523
97      229
98       77
99       33
100      19
Name: points, Length: 22, dtype: int64

groupby() created a group of reviews which allotted the same point values to the given wines. Then, for each of these groups, we grabbed the points() column and counted how many times it appeared.  value_counts() is just a shortcut to this groupby() operation.
  
We can use any of the summary functions we've used before with this data. For example, to get the cheapest wine in each point value category, we can do the following:

In [24]:
reviews.groupby('points').price.mean()

points
60      15.000000
80      16.372152
81      17.182353
82      18.870767
83      18.237353
          ...    
96     159.292531
97     207.173913
98     245.492754
99     284.214286
100    485.947368
Name: price, Length: 22, dtype: float64

You can think of each group we generate as being a slice of our DataFrame containing only data with values that match. This DataFrame is accessible to us directly using the apply() method, and we can then manipulate the data in any way we see fit. 

For example, here's one way of selecting the name of the first wine reviewed from each winery in the dataset:

In [33]:
reviews.groupby('winery').apply(lambda df: df.title.iloc[0])

winery
1+1=3                                     1+1=3 NV Rosé Sparkling (Cava)
10 Knots                            10 Knots 2010 Viognier (Paso Robles)
100 Percent Wine              100 Percent Wine 2015 Moscato (California)
1000 Stories           1000 Stories 2013 Bourbon Barrel Aged Zinfande...
1070 Green                  1070 Green 2011 Sauvignon Blanc (Rutherford)
                                             ...                        
Órale                       Órale 2011 Cabronita Red (Santa Ynez Valley)
Öko                    Öko 2013 Made With Organically Grown Grapes Ma...
Ökonomierat Rebholz    Ökonomierat Rebholz 2007 Von Rotliegenden Spät...
àMaurice               àMaurice 2013 Fred Estate Syrah (Walla Walla V...
Štoka                                    Štoka 2009 Izbrani Teran (Kras)
Length: 16757, dtype: object

For even more fine-grained control, you can also group by more than one column.

For an example, here's how we would pick out the best wine by country and province:

In [40]:
reviews.groupby(['country', 'province']).apply(lambda df: df.loc[df.points.idxmax()])

country   
country   province                      
Argentina Mendoza Province  Argentina  \
          Other             Argentina   
Armenia   Armenia             Armenia   
Australia Australia Other   Australia   
          New South Wales   Australia   
...                               ...   
Uruguay   Juanico             Uruguay   
          Montevideo          Uruguay   
          Progreso            Uruguay   
          San Jose            Uruguay   
          Uruguay             Uruguay   

                                                                  description   
country   province                                                              
Argentina Mendoza Province  If the color doesn't tell the full story, the ...  \
          Other             Take note, this could be the best wine Colomé ...   
Armenia   Armenia           Deep salmon in color, this wine offers a bouqu...   
Australia Australia Other   Writes the book on how to make a wine filled w...   
          New South Wales   De Bortoli's Noble One is as good as ever in 2...   
...                                                                       ...   
Uruguay   Juanico           This mature Bordeaux-style blend is earthy on ...   
          Montevideo        A rich, heady bouquet offers aromas of blackbe...   
          Progreso          Rusty in color but deep and complex in nature,...   
          San Jose          Baked, sweet, heavy aromas turn earthy with ti...   
          Uruguay           Cherry and berry aromas are ripe, healthy and ...   

                                                        designation  points   
country   province                                                            
Argentina Mendoza Province                         Nicasia Vineyard      97  \
          Other                                             Reserva      95   
Armenia   Armenia                                    Estate Bottled      88   
Australia Australia Other                             Sarah's Blend      93   
          New South Wales                        Noble One Bortytis      94   
...                                                             ...     ...   
Uruguay   Juanico                  Preludio Barrel Select Lote N 77      90   
          Montevideo        Monte Vide Eu Tannat-Merlot-Tempranillo      91   
          Progreso                   Etxe Oneko Fortified Sweet Red      90   
          San Jose                         El Preciado Gran Reserva      87   
          Uruguay                         Blend 002 Limited Edition      91   

                            price          province                 region_1   
country   province                                                             
Argentina Mendoza Province  120.0  Mendoza Province                  Mendoza  \
          Other              90.0             Other                    Salta   
Armenia   Armenia            15.0           Armenia                      NaN   
Australia Australia Other    15.0   Australia Other  South Eastern Australia   
          New South Wales    32.0   New South Wales          New South Wales   
...                           ...               ...                      ...   
Uruguay   Juanico            45.0           Juanico                      NaN   
          Montevideo         60.0        Montevideo                      NaN   
          Progreso           46.0          Progreso                      NaN   
          San Jose           50.0          San Jose                      NaN   
          Uruguay            22.0           Uruguay                      NaN   

                           region_2        taster_name taster_twitter_handle   
country   province                                                             
Argentina Mendoza Province      NaN  Michael Schachner           @wineschach  \
          Other                 NaN  Michael Schachner           @wineschach   
Armenia   Armenia               NaN      Mike DeSimon

Another groupby() method worth mentioning is agg(), which lets you run a bunch of different functions on your DataFrame simultaneously. 

For example, we can generate a simple statistical summary of the dataset as follows:

In [45]:
reviews.groupby(['country', 'province']).price.agg([len, min, max])

len   min    max
country   province                           
Argentina Mendoza Province  3264   4.0  230.0
          Other              536   7.0  150.0
Armenia   Armenia              2  14.0   15.0
Australia Australia Other    245   5.0  130.0
          New South Wales     85   8.0  125.0
...                          ...   ...    ...
Uruguay   Juanico             12  10.0  130.0
          Montevideo          11  17.0   60.0
          Progreso            11  12.0   46.0
          San Jose             3  20.0   50.0
          Uruguay             24  12.0   50.0

[425 rows x 3 columns]

**Note :** Effective use of groupby() will allow you to do lots of really powerful things with your dataset

### Multi-indexes

In all of the examples we've seen thus far we've been working with DataFrame or Series objects with a single-label index. groupby() is slightly different in the fact that, depending on the operation we run, it will sometimes result in what is called a multi-index.

A multi-index differs from a regular index in that it has multiple levels. 

For example:

In [46]:
countries_reviewed = reviews.groupby(['country', 'province']).description.agg([len])
countries_reviewed

len
country   province              
Argentina Mendoza Province  3264
          Other              536
Armenia   Armenia              2
Australia Australia Other    245
          New South Wales     85
...                          ...
Uruguay   Juanico             12
          Montevideo          11
          Progreso            11
          San Jose             3
          Uruguay             24

[425 rows x 1 columns]

In [47]:
mi = countries_reviewed.index
type(mi)
mi

MultiIndex([('Argentina',  'Mendoza Province'),
            ('Argentina',             'Other'),
            (  'Armenia',           'Armenia'),
            ('Australia',   'Australia Other'),
            ('Australia',   'New South Wales'),
            ('Australia',   'South Australia'),
            ('Australia',          'Tasmania'),
            ('Australia',          'Victoria'),
            ('Australia', 'Western Australia'),
            (  'Austria',           'Austria'),
            ...
            (       'US',        'Washington'),
            (       'US', 'Washington-Oregon'),
            (  'Ukraine',           'Ukraine'),
            (  'Uruguay',         'Atlantida'),
            (  'Uruguay',         'Canelones'),
            (  'Uruguay',           'Juanico'),
            (  'Uruguay',        'Montevideo'),
            (  'Uruguay',          'Progreso'),
            (  'Uruguay',          'San Jose'),
            (  'Uruguay',           'Uruguay')],
           names=['coun

Multi-indices have several methods for dealing with their tiered structure which are absent for single-level indices. They also require two levels of labels to retrieve a value. Dealing with multi-index output is a common "gotcha" for users new to pandas.

The use cases for a multi-index are detailed alongside instructions on using them in the MultiIndex / Advanced Selection section of the pandas documentation.

However, in general the multi-index method you will use most often is the one for converting back to a regular index, the reset_index() method:

In [48]:
countries_reviewed.reset_index()

country          province   len
0    Argentina  Mendoza Province  3264
1    Argentina             Other   536
2      Armenia           Armenia     2
3    Australia   Australia Other   245
4    Australia   New South Wales    85
..         ...               ...   ...
420    Uruguay           Juanico    12
421    Uruguay        Montevideo    11
422    Uruguay          Progreso    11
423    Uruguay          San Jose     3
424    Uruguay           Uruguay    24

[425 rows x 3 columns]

### Sorting

Looking again at countries_reviewed we can see that grouping returns data in index order, not in value order. That is to say, when outputting the result of a groupby, the order of the rows is dependent on the values in the index, not in the data.

To get data in the order want it in we can sort it ourselves. The sort_values() method is handy for this.

In [51]:
countries_reviewed = countries_reviewed.reset_index()
countries_reviewed.sort_values(by='len')

level_0  index       country               province    len
179      179    179        Greece  Muscat of Kefallonian      1
192      192    192        Greece          Sterea Ellada      1
194      194    194        Greece                 Thraki      1
354      354    354  South Africa             Paardeberg      1
40        40     40        Brazil       Serra do Sudeste      1
..       ...    ...           ...                    ...    ...
409      409    409            US                 Oregon   5373
227      227    227         Italy                Tuscany   5897
118      118    118        France               Bordeaux   5941
415      415    415            US             Washington   8639
392      392    392            US             California  36247

[425 rows x 5 columns]

sort_values() defaults to an ascending sort, where the lowest values go first. However, most of the time we want a descending sort, where the higher numbers go first. That goes thusly:

In [52]:
countries_reviewed.sort_values(by='len', ascending=False)

level_0  index       country         province    len
392      392    392            US       California  36247
415      415    415            US       Washington   8639
118      118    118        France         Bordeaux   5941
227      227    227         Italy          Tuscany   5897
409      409    409            US           Oregon   5373
..       ...    ...           ...              ...    ...
101      101    101       Croatia              Krk      1
247      247    247   New Zealand        Gladstone      1
357      357    357  South Africa  Piekenierskloof      1
63        63     63         Chile          Coelemu      1
149      149    149        Greece           Beotia      1

[425 rows x 5 columns]

To sort by index values, use the companion method sort_index(). This method has the same arguments and default order:

In [53]:
countries_reviewed.sort_index()

level_0  index    country          province   len
0          0      0  Argentina  Mendoza Province  3264
1          1      1  Argentina             Other   536
2          2      2    Armenia           Armenia     2
3          3      3  Australia   Australia Other   245
4          4      4  Australia   New South Wales    85
..       ...    ...        ...               ...   ...
420      420    420    Uruguay           Juanico    12
421      421    421    Uruguay        Montevideo    11
422      422    422    Uruguay          Progreso    11
423      423    423    Uruguay          San Jose     3
424      424    424    Uruguay           Uruguay    24

[425 rows x 5 columns]

Finally, know that you can sort by more than one column at a time:

In [54]:
countries_reviewed.sort_values(by=['country', 'len'])

level_0  index    country          province   len
1          1      1  Argentina             Other   536
0          0      0  Argentina  Mendoza Province  3264
2          2      2    Armenia           Armenia     2
6          6      6  Australia          Tasmania    42
4          4      4  Australia   New South Wales    85
..       ...    ...        ...               ...   ...
421      421    421    Uruguay        Montevideo    11
422      422    422    Uruguay          Progreso    11
420      420    420    Uruguay           Juanico    12
424      424    424    Uruguay           Uruguay    24
419      419    419    Uruguay         Canelones    43

[425 rows x 5 columns]

In [55]:
countries_reviewed.sort_values(by=['country', 'len'], ascending=[False, True])

level_0  index    country          province   len
423      423    423    Uruguay          San Jose     3
418      418    418    Uruguay         Atlantida     5
421      421    421    Uruguay        Montevideo    11
422      422    422    Uruguay          Progreso    11
420      420    420    Uruguay           Juanico    12
..       ...    ...        ...               ...   ...
7          7      7  Australia          Victoria   322
5          5      5  Australia   South Australia  1349
2          2      2    Armenia           Armenia     2
1          1      1  Argentina             Other   536
0          0      0  Argentina  Mendoza Province  3264

[425 rows x 5 columns]

In [58]:
reviews.sort_values(by=["price"], ascending=False).iloc[0,:].price

3300.0

## Pandas - Data types and missing values
### Introduction

Pandas can works with differents data types and it will create new data types. We'll also see how we can replace and deal with missing values.

### Dtypes

The data type for a column in a DataFrame or a Series is known as the dtype.

You can use the dtype property to grab the type of a specific column. For instance, we can get the dtype of the price column in the reviews DataFrame:

In [ ]:
reviews.price.dtype

Alternatively, the dtypes property returns the dtype of every column in the DataFrame:

In [ ]:
reviews.dtypes

Data types tell us something about how pandas is storing the data internally. float64 means that it's using a 64-bit floating point number; int64 means a similarly sized integer instead, and so on.

One peculiarity to keep in mind (and on display very clearly here) is that columns consisting entirely of strings do not get their own type; they are instead given the object type.

It's possible to convert a column of one type into another wherever such a conversion makes sense by using the astype() function. 

For example, we may transform the points column from its existing int64 data type into a float64 data type:

In [ ]:
print(reviews.points.dtypes)
reviews.points.astype('float64').astype('int64')

A DataFrame or Series index has its own dtype, too:

In [ ]:
reviews.index.dtype

### Missing data

Entries missing values are given the value NaN, short for "Not a Number". For technical reasons these NaN values are always of the float64 dtype.

Pandas provides some methods specific to missing data. To select NaN entries you can use pd.isnull() (or its companion pd.notnull()). This is meant to be used thusly:

In [ ]:
reviews[pd.isnull(reviews.country)]

Replacing missing values is a common operation. Pandas provides a really handy method for this problem: fillna(). fillna() provides a few different strategies for mitigating such data. For example, we can simply replace each NaN with an "Unknown":

In [ ]:
reviews.region_2.fillna("Unknown")

Or we could fill each missing value with the first non-null value that appears sometime after the given record in the database. This is known as the backfill strategy.

Alternatively, we may have a non-null value that we would like to replace. For example, suppose that since this dataset was published, reviewer Kerin O'Keefe has changed her Twitter handle from @kerinokeefe to @kerino. One way to reflect this in the dataset is using the replace() method:

In [ ]:
reviews.taster_twitter_handle.replace("@kerinokeefe", "@kerino")

The replace() method is worth mentioning here because it's handy for replacing missing data which is given some kind of sentinel value in the dataset: things like "Unknown", "Undisclosed", "Invalid", and so on.

We can also choose to not use a row that has missing data by using the dropna() method:

In [ ]:
reviews.dropna(inplace=True)

the dropna() method will remove any row that has missing data.
the inplace parameter will modify the original DataFrame rather than returning a new one.

## Pandas - renamming and combining
### Introduction

Oftentimes data will come to us with column names, index names, or other naming conventions that we are not satisfied with. In that case, you'll learn how to use pandas functions to change the names of the offending entries to something better.

You'll also explore how to combine data from multiple DataFrames and/or Series.

### Renaming

The first function we'll introduce here is rename(), which lets you change index names and/or column names. 

For example, to change the points column in our dataset to score, we would do:

In [ ]:
reviews.rename(columns={'points': 'score'}, inplace=True)

rename() lets you rename index or column values by specifying a index or column keyword parameter, respectively. It supports a variety of input formats, but usually a Python dictionary is the most convenient. Here is an example using it to rename some elements of the index.

In [ ]:
reviews.rename(index={4: 'firstEntry', 10: 'secondEntry'})

You'll probably rename columns very often, but rename index values very rarely. For that, set_index() is usually more convenient.

Both the row index and the column index can have their own name attribute. The complimentary rename_axis() method may be used to change these names. For example:

In [ ]:
reviews.rename_axis("wines", axis='rows').rename_axis("fields", axis='columns')

### Combining

When performing operations on a dataset, we will sometimes need to combine different DataFrames and/or Series in non-trivial ways. Pandas has three core methods for doing this. In order of increasing complexity, these are concat(), join(), and merge(). Most of what merge() can do can also be done more simply with join(), so we will omit it and focus on the first two functions here.

The simplest combining method is concat(). Given a list of elements, this function will smush those elements together along an axis.

This is useful when we have data in different DataFrame or Series objects but having the same fields (columns). One example: the YouTube Videos dataset, which splits the data up based on country of origin (e.g. Canada and the UK, in this example). If we want to study multiple countries simultaneously, we can use concat() to smush them together:

In [ ]:
canadian_youtube = pd.read_csv("./data/numpy_pandas/dataset/youtube/CAvideos.csv")
print(canadian_youtube.shape)
british_youtube = pd.read_csv("./data/numpy_pandas/dataset/youtube/GBvideos.csv")
print(british_youtube.shape)

pd.concat([canadian_youtube, british_youtube])

The middlemost combiner in terms of complexity is join(). join() lets you combine different DataFrame objects which have an index in common. For example, to pull down videos that happened to be trending on the same day in both Canada and the UK, we could do the following:

In [ ]:
left = canadian_youtube.set_index(['title', 'trending_date'])
right = british_youtube.set_index(['title', 'trending_date'])

left.join(right, lsuffix='_CAN', rsuffix='_UK')

Here we do a join on two DataFrames based on the index of the two DataFrames. The lsuffix and rsuffix parameters are used to indicate the suffixes of the column names in the left and right DataFrame, respectively.

The lsuffix and rsuffix parameters are necessary here because the data has the same column names in both British and Canadian datasets. If this wasn't true (because, say, we'd renamed them beforehand) we wouldn't need them.

In [ ]:
np.random.seed()
ts = pd.Series(np.random.randn(1000), index=pd.date_range("1/1/2000", periods=1000))
ts = ts.cumsum()
ts.plot();